python/ffsim/tenpy/hamiltonians/molecular_hamiltonian.py

```diff
@@ -0,0 +1,149 @@
+# that they have been altered from the originals.
+
+"""TeNPy molecular Hamiltonian."""
+
+from __future__ import annotations
+
+import itertools
+
+import numpy as np
+from tenpy.models.lattice import Lattice
+from tenpy.models.model import CouplingMPOModel
+from tenpy.networks.site import SpinHalfFermionSite
+
+from ffsim.hamiltonians.molecular_hamiltonian import MolecularHamiltonian
+
+# ignore lowercase variable checks to maintain TeNPy naming conventions
+# ruff: noqa: N806
+
+
+class MolecularHamiltonianMPOModel(CouplingMPOModel):
+    """Molecular Hamiltonian."""
+
+    def init_sites(self, params) -> SpinHalfFermionSite:
+        """Initialize sites."""
+        return SpinHalfFermionSite()
+
+    def init_lattice(self, params) -> Lattice:
+        """Initialize lattice."""
+        one_body_tensor = params.get("one_body_tensor", None, expect_type="array")
+        norb = one_body_tensor.shape[0]
+        site = self.init_sites(params)
+        basis = np.array(([norb, 0], [0, 1]))
+        pos = np.array([[i, 0] for i in range(norb)])
+        lat = Lattice(
+            [1, 1],
+            [site] * norb,
+            basis=basis,
+            positions=pos,
+        )
+        return lat
+
+    def init_terms(self, params) -> None:
+        """Initialize terms."""
+        dx0 = np.array([0, 0])
+        one_body_tensor = params.get("one_body_tensor", None, expect_type="array")
+        two_body_tensor = params.get("two_body_tensor", None, expect_type="array")
+        constant = params.get("constant", 0, expect_type="real")
+        norb = one_body_tensor.shape[0]
+
+        for p in range(norb):
+            h1 = one_body_tensor[p, p]
+            self.add_onsite(h1, p, "Ntot")
+            h2 = two_body_tensor[p, p, p, p]
+            self.add_onsite(h2, p, "Ntot")
+            self.add_onsite(-0.5 * h2, p, "Nu Nu")
+            self.add_onsite(-0.5 * h2, p, "Cdu Cd Cdd Cu")
+            self.add_onsite(-0.5 * h2, p, "Cdd Cu Cdu Cd")
+            self.add_onsite(-0.5 * h2, p, "Nd Nd")
+            self.add_onsite(constant / norb, p, "Id")
+
+        for p, q in itertools.combinations(range(norb), 2):
+            self.add_coupling(
+                one_body_tensor[p, q], q, "Cdu", p, "Cu", dx0, plus_hc=True
+            )
+            self.add_coupling(
+                one_body_tensor[p, q], q, "Cdd", p, "Cd", dx0, plus_hc=True
+            )
+
+        for p, s in itertools.combinations_with_replacement(range(norb), 2):
+            for q, r in itertools.combinations_with_replacement(range(norb), 2):
+                if not p == q == r == s:
+                    indices = [(p, q, r, s)]
+                    if p < s:
+                        indices.append((s, q, r, p))
+                    if q < r:
+                        indices.append((p, r, q, s))
+
+                    for idx, (i, j, k, ell) in enumerate(indices):
+                        flag_hc = True if not idx and i < ell and j < k else False
```

Comment:
This logic is somewhat convoluted, would it be simplified if we handled the case of two unique indices in the preceding
```
for p, q in itertools.combinations(range(norb), 2):
```
loop? In that loop you currently only handle one-body terms.

Reply:
One possibility that could make things more readable, is if we do not make use of the `plus_hc` keyword argument of the `add_multi_coupling` method. In this case, I could simply add an additional `if` statement:

```
if p<s and q<r:
    indices.append((s, r, q, p))
```

Reply:
I still like my idea of handling some of the two-body terms in the
```python
for p, q in itertools.combinations(range(norb), 2):
```
loop, since that loop currently only handles one-body terms.